File: cpgateway/app/models/org.py
import uuid
from enum import IntEnum
from sqlalchemy import Column, Integer, BigInteger, String, DateTime, ForeignKey
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from app.core.database import Base


class OrgType(IntEnum):
    TEAM = 1
    SYSTEM = 2


class OrgStatus(IntEnum):
    PENDING   = 0  # 待激活（用户注册后未激活账号）
    ACTIVE    = 1  # 正常运营
    SUSPENDED = 2  # 暂停（资源保留，禁止写操作）
    DISABLED  = 3  # 禁用（禁止登录和任何访问）


class Organization(Base):
    """
    全局组织模型 (Organization)
    Org 是全局概念，跨所有 Region 共享。资源通过 org_id 归属到 Org。

    字段说明:
    - slug: 唯一可读标识，创建后不可修改，用于 URL 友好引用。
    - org_type: TEAM(1) 普通团队, SYSTEM(2) 系统级 Org（全局唯一，不可删除）。
    - owner_user_id: Org 的 Owner，拥有最高管理权。
    - default_sg: 默认安全组 ID（由 Cloudland 创建后回填）。
    """
    __tablename__ = "organizations"

    id = Column(BigInteger, primary_key=True, index=True, autoincrement=True)
    uuid = Column(String(36), unique=True, index=True, default=lambda: str(uuid.uuid4()))
    name = Column(String(255), nullable=False)
    slug = Column(String(64), unique=True, index=True, nullable=False)
    org_type = Column(Integer, default=OrgType.TEAM, nullable=False)
    owner_user_id = Column(BigInteger, ForeignKey("users.id"), nullable=False)
    default_sg = Column(BigInteger, default=0)
    status = Column(Integer, default=OrgStatus.PENDING, nullable=False)

    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    deleted_at = Column(DateTime(timezone=True), nullable=True)

    # Relationships
    owner = relationship("User", foreign_keys=[owner_user_id])
    members = relationship("Member", back_populates="organization", cascade="all, delete-orphan")
    resource_quotas = relationship("OrgResourceQuota", back_populates="org", cascade="all, delete-orphan")
    resource_consumptions = relationship("OrgResourceConsumption", back_populates="org", cascade="all, delete-orphan")

File: cpgateway-go/src/model/org.go
package model

import (
	"cpgateway-go/src/dbs"

	"gorm.io/gorm"
)

type OrgType int

const (
	OrgTeam   OrgType = 1
	OrgSystem OrgType = 2
)

type OrgStatus int

const (
	OrgPending   OrgStatus = 0
	OrgActive    OrgStatus = 1
	OrgSuspended OrgStatus = 2
	OrgDisabled  OrgStatus = 3
)

type Organization struct {
	Model
	Name        string    `gorm:"type:varchar(255);not null" json:"name"`
	Slug        string    `gorm:"type:varchar(64);not null" json:"slug"`
	Description string    `gorm:"type:text;default:''" json:"description"`
	OrgType     OrgType   `gorm:"default:1;not null" json:"org_type"`
	OwnerUserID int64     `gorm:"not null" json:"owner_user_id"`
	DefaultSG   int64     `gorm:"default:0" json:"default_sg"`
	Status      OrgStatus `gorm:"default:0;not null" json:"status"`

	// Relationships (not stored in DB)
	Owner   *User    `gorm:"foreignKey:OwnerUserID" json:"owner,omitempty"`
	Members []Member `gorm:"foreignKey:OrgID" json:"members,omitempty"`
}

func (Organization) TableName() string {
	return "organizations"
}

func init() {
	dbs.AutoMigrate(&Organization{})
	dbs.AutoUpgrade("003_org_slug_unique", func(db *gorm.DB) error {
		return db.Exec(`
			CREATE UNIQUE INDEX IF NOT EXISTS uq_org_slug_active
			ON organizations (slug)
			WHERE deleted_at IS NULL
		`).Error
	})
}
